src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,407 @@
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method creates a ConstraintSerializer, which can deliver the
+     * constraint/slot to the real serializer. If a costume serialization logic
+     * is needed, user can have a subclass of ConstraintSerializer, override the
+     * serialize method with the costume logic, and then override this method
+     * with returning the instance of new subclass.
+     * 
+     * @param backEndType
+     * @param lattice
+     * @return A deliverer Serializer.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+        return new ConstraintSerializer<>(backEndType, lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType.fullyQualifiedName + "BackEnd");
```

Comment:
Using backEnd.class in `BackEndType` instead of the string sounds great, and we can get rid of the if statement in `ConstraintSerializer` through it. However, I think we may want to keep `ConstraintSerializer` and still use `GeneralSolver` to create back end and serializer. As I mentioned yesterday, with this deliverer logic, we can easily to implement a custom serialization logic.

For example, for type system X with combine constraint, when we use Max-SAT back end to solve its constraints, we wanted to implement its custom serialization for combine constraint since there is no general serialization for combine constraint in `MaxSatSerializer`. So what we are gonna to do is create a subclass of `ConstraintSerializer` and override  `serialize(CombineConstraint combineConstraint)`
```java
public class XConstraintSerializer extends ConstraintSerializer {
    @Override
    public T serialize(CombineConstraint combineConstraint) {
        //implement the serialization logic for type system X.
    }
}
```

Then, like what we did for Dataflow and Ontology, create a `XGeneralSolver.class`, and override `createSerializer` method.
```java
public class DataflowGeneralSolver extends GeneralSolver {
    @Override
    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
        return new XConstraintSerializer();
    }
}
```
Such that we create a solver for type system X in a relative convenient way.